bookdown/03-optimization.Rmd

```diff
@@ -231,6 +231,171 @@ Note that we do not expect any differences here compared to the non-tuned approa
+The arising tradeoff, that has to be made, is called the "$n$ versus $B/n$" problem.
+To create a balance between exploitation by selection based on early training performance versus exploration of training performances in later training stages, several brackets are constructed with an associated set of configurations of varying sizes.
+Different brackets are initialized with different number of configurations, and different initial budget sizes for the first stage, but each bracket is assigned (approximately) the same budget.
+Some brackets contain many configurations, with a small initial budget, so that many are discarded after having been trained for only a short amount of time.
+Others are constructed with few configurations, where discarding only takes place after a significant amount of budget was consumed.
+The former bracket corresponds to the exploitation (or high $n$), while the latter to the exploration (or high $B/n$).
+
+The configurations at the start of each bracket are initialized by stochastic, often uniform, sampling.
+Each bracket is divided into multiple stages, with a stage acting as a termination judge.
+The first stage allocates a small budget (the more configurations to start with, the smaller the budget) to each configuration to then discard the worst performing ones, after the budget is consumed.
+The fraction $1/\eta$ of discarded configurations in each stage is given by the user in the form of a numeric argument eta bigger than `1`.
+The following stage then takes the survivors and allocates additional budget to them (less survivors means more budget for each), and the survival game starts again.
+Because the whole budget of a bracket is equally split up between the stages, the budget of a single (surviving) configuration is increased by a factor of eta between follow-up stages.
+This runs until the budget of a bracket is consumed.
+To end up with only a small amount of survivors after the last stage terminates, the amount of configurations to start with is calculated at the start of each bracket.
+Note that currently all configurations are trained completely from the beginning, so no online updates of models from stage to stage is happening.
+
+To identify the budget for evaluating hyperband, the user has to specify explicitly which hyperparameter of the learner influences the budget by extending a single hyperparameter in the `r ref("ParamSet")` with an argument (`tags = "budget"`), like in the following snippet:
+
+```{r 03-optimization-012}
+library(paradox)
+
+# Hyperparameter subset of XGBoost
+params = list(
+  ParamInt$new("nrounds", lower = 1, upper = 16, tags = "budget"),
+  ParamFct$new("booster", levels = c("gbtree", "gblinear", "dart"))
+)
+```
+
+Thanks to the broad ecosystem of the `r mlr_pkg("mlr3verse")` a learner does not require a natural budget parameter.
```

Comment:
how is not requiring a natural budget parameter a consequence of the "broad ecosystem"?